Fix issue with linked list remove

btheado opened the pull request:
Prior to this fix, the filter expression `A =A B =A :filter[match[B]]` resulted in `B A` instead of just `B`.

The prev pointer of the underlying double linked-list implementation was getting corrupted for this case. The code for `prev` was "symmetric" with the `next` code, even though for `prev` there can be an extra `undefined`  at the front of the list, but the same extra `undefined` entry is not present at the end of `next`.

Fixes #7059

Jermolene commented:
Thanks @btheado. @flibbles are you available to review this?

flibbles commented:
I tested it. It looks good. Sorry you guys had to dig through the linked-list yourselves. I know it's super logic heavy. This fix does seem to resolve the issue and remove the special exception which apparently is not needed (not sure how that happened.)

> Ouch thank you @btheado. The linked list code is quite challenging to work with, might you be able to investigate a PR?

@Jermolene, I have a refactor of the link-list class (https://github.com/Jermolene/TiddlyWiki5/pull/6056) which I still promote. It's a lot easier to understand, and it's a bit faster if I remember, and it doesn't suffer from the same bug (even if it had the same strange exception...)

Once this is merged, I'd be happy to fix any conflicts and re-prep my refactor.

btheado commented:
Thanks, @flibbles.

As mentioned in #7059, my initial fix was to just comment out the original line 141. It didn't seem right to me, but I was having trouble figuring out the code to know for sure.  Then I added some code to linked-list which would start at the end of the list and accumulate the items by following the prev links and then reverse the list once finished. I added that to the `compare` function in the tests. That way the arrays generated for comparison were fully using both the next and prev links.

Sure enough with this extra testing of the prev links and with my original fix some of the tests failed. However, with my final fix all these tests passed. This gives me more confidence in the fix.

I didn't think it would make sense to include this extra linked-list code I wrote, so I didn't. However, it occurs to me it should still be possible to write SOME tests using the existing linked-list methods which make more use of the prev links would have exposed that issue. I'll think about that some more.

flibbles commented:
I would think if a reverse iteration was revealing some corruption of the list, then it's worth having, even if it's not worth having in the link-list module itself. The test-linked-list file could inject it into LinkList.prototype and test there. That's just if we really don't want to add it, but I can think if instances where reverse iteration may be desired in the rest of core, so maybe add it?

Jermolene commented:
Thanks @btheado @flibbles
